### PR TITLE
New Command 'core:invalidate-report-data' to invalidate archive data (w/ period cascading)

### LINKED ISSUE #8858: invalidateArchivedReports API to invalidate report data for a particular segment
In: https://piwik.org/faq/how-to/faq_155/

A typical API query to invalidate reports is: `?module=API&method=CoreAdminHome.invalidateArchivedReports&idSites=1,3,5&dates=2012-01-01,2011-10-15&token_auth=xyz`

It would be useful when managing a Piwik instance, to be able to invalidate report data for a particular segment only. Why? eg. when invalidating data for a very long date range, sometimes we want to invalidate a particular segment data that we care most about, there is no need to have all other report data reprocessed ( sometimes this could take several days to compute).

Steps

 * Implement feature
 * Include regression test
 * [Update FAQ](https://piwik.org/faq/how-to/faq_155/) to document this new feature

### PR DESCRIPTION
**Pull request by diosmosis**
This PR contains a new command `core:invalidate-report-data` that provides a user friendly way to invalidate archive data (as an alternative to using the API multiple times).

The command supports invalidating data for multiple sites, dates & periods and also supports cascading periods, so all child periods will be invalidated to. The command is smart enough to expand the range for child periods, so if a month is requested for invalidation, but it's first week has some days in the previous month, then those days will also be invalidated.

Also fixes #8858
## TODO
- [x] Integration test for the command.
- [x] Merge command logic w/ API logic + add redundant tests for API.
- [x] Add upwards cascading as default behavior + tests. Clean up ArchiveInvalidator in process.
- [x] Rewrite documentation for invalidation method.
- [x] allow invalidating for a segment (allow multiple in command)

### CODE REVIEW

**Comment by tsteur** [comment]
Haven't tested it but LGTM. My usual comment: Would be nice to extract some private methods into other classes making them public and better (independent) testable but up to you :)

**Comment by diosmosis** [comment]
> Would be nice to extract some private methods into other classes making them public and better (independent) testable but up to you

Perhaps you could be more specific? If you have to repeat yourself, it means your method of communication isn't having the desired effect.

One method is public in the command and is tested in a unit test. The others are only for processing input options. I don't know where to put those where they wouldn't just create clutter. Is there some symfony command way of defining those (ie, something like the DialogHelper) I am not aware of?

**Comment by tsteur** [comment]
I didn't mean the input ones sorry should have mentioned it in the code. Will leave some comments. I'm just personally trying to have as many public methods as possible so the methods are testable and reusable - if needed. Having one or two more classes should rather remove clutter from one big class instead of adding new clutter.

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        /** @var Date $realEndDate */
+        $realEndDate = end($subperiods)->getDateEnd();
+        if ($realEndDate->isLater(Date::factory($endDate))) {
+            $endDate = $realEndDate->toString();
+        }
+
+        $this->addPeriodsToCascadeOn($result, $childPeriodType, $startDate, $endDate);
+    }
+
+    private function getChildPeriod($periodType)
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: Maybe this could be moved into a period class? A `switch` is most times a sign that one needs subclasses for each entity. So each period class could have a method `getChildPeriod()`. So you get the Period class for your period via the factory and then call `$period->getChildPeriod()`

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+    /**
+     * TODO
+     *
+     * @param Date[][] $result
+     * @param string $periodType
+     * @param string $startDate
+     * @param string $endDate
+     */
+    private function addPeriodsToCascadeOn(&$result, $periodType, $startDate, $endDate)
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: Maybe this could be moved out of the command and tested separately?

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        return $result;
+    }
+
+    /**
+     * @param string[] $periodTypes
+     * @param Date[][] $dateRanges
+     * @param bool $cascade
+     * @return Date[][]
+     */
+    public function getPeriodsToInvalidateFor($periodTypes, $dateRanges, $cascade)
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: This one is public although not needed to have it public here right? I think it's public so that it can be tested so maybe it can be moved in another class as well where it is public and maybe "needed" to be public

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return Date[][]
+     */
+    private function getDateRangesToInvalidateFor(InputInterface $input)
+    {
+        $dateRanges = $input->getOption('dates');
+        if (empty($dateRanges)) {
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: The code starting from here could be kinda moved into a public method in another class as well. It would be could to have separate, small tests for this one I think as I would for example not feel confident to change it without "fear" or a certain "risk". I know it's tested by the command itself but not sure if really each case it goes is tested

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            case 'week':
+                return 'day';
+            default:
+                return null;
+        }
+    }
+
+    private function validateDate($dateString)
+    {
+        try {
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: Maybe here we could have something like `Date::validate()` or so? I think it would be useful to be able to reuse this one and to write a simple test for it

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                }
+            }
+        }
+    }
+
+    private function getSitesToInvalidateFor(InputInterface $input)
+    {
+        $sites = $input->getOption('sites');
+
+        $siteIds = Site::getIdSitesFromIdSitesString($sites);
```
Comment (tsteur) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: I think we have same code eg in `Access::checkUserHasAdminAccess` and possibly many other parts. Could be nice to extract it or maybe there is something that can be reused already

File: plugins/CoreAdminHome/Commands/InvalidateReportData.php (hunk 5 above)
Comment (diosmosis) [inline plugins/CoreAdminHome/Commands/InvalidateReportData.php]: Well, it's technically just `Date::factory()`, so it wouldn't need much testing. The method is there so its use is more self-explanatory.

**Comment by diosmosis** [comment]
@tsteur Ok, I understand what you mean now. Some of your suggestions I can look into with this PR, others are unrelated. Also I don't consider them to be high value work, so I will probably create issues for them in the 3.0 branch

I think most of these suggestions have to do w/ handling console input, really. Eg, getting & validating date ranges + idsites, etc. Perhaps we can create a custom symfony console helper for reuse.

**Comment by diosmosis** [comment]
@tsteur Created an issue here: https://github.com/piwik/piwik/issues/8844

**Comment by tsteur** [comment]
I think not all are console related. Eg the idSites check can be used in various scenarios but a console helper will be certainly nice :+1:

**Comment by diosmosis** [comment]
> I think not all are console related.

Mentioned this at the end of the issue I created (when I realized this ;)). I'm not sure what a more "universal" concept would be though. Anyway not something I will work on now :)

**Comment by diosmosis** [comment]
@tsteur Moved some code into other places and added more tests, can you check that you see nothing wrong w/ the last couple commits?

**Comment by mattab** [comment]
>  a new command core:invalidate-report-data that provides a user friendly way to invalidate archive data (as an alternative to using the API multiple times).

this command is useful, and it would be awesome to push the functionality down to the API. Can we enhance existing API call (documented in this FAQ and commonly used), or do we need to create new API? then it would make sense to even validate the inputs in API and move logic (eg. friendly error messages) from command to API. what do you think?

**Comment by mattab** [comment]
Looking at the code there is a bug: when invalidating a day, it currently doesnt invalidate week/month/year but should (since one of the day has changed).

As per discussion:

- we need "upwards cascading" as default behavior (for data correctness) 
- we need downwards cascading, on-demand (eg. via new invalidateArchivedReports API  parameter `cascade=1`) 
   - we can move more command logic to the API itself (bonus: many use this API already, and CLI access is in general is more difficult)

**Comment by quba** [comment]
@mattab what about invalidating single segment?

**Comment by mattab** [comment]
@quba sounds good to me...eg. `&segment=pageUrlxxyzz` would invalidate only archives for this segment. Would be especially easy to use when we do #8209 then we could simply use `&segment=idSegment==6` or so

@diosmosis do you maybe want to look at adding `segment` param along with other changes?

**Comment by mattab** [comment]
> @diosmosis do you maybe want to look at adding segment param along with other changes?

Covered in #8858

**Comment by diosmosis** [comment]
Ready for another review. Scope was increased quite a bit, so needs a extensive review.

**Comment by mattab** [comment]
Todo:

* [ ] Review & Merge
* [ ] Update FAQ: https://piwik.org/faq/how-to/faq_155/

**Comment by mattab** [comment]
related:  New plugin: provide simple UI to invalidate old reports #8942

File: core/Archive/ArchiveInvalidator.php [hunk 7]
```diff
@@ -120,17 +126,28 @@ public function forgetRememberedArchivedReportsToInvalidate($idSite, Date $date)
...
+     * @param $segment Segment
+     * @param bool $cascadeDown
+     * @return InvalidationResultInfo
      * @throws \Exception
      */
-    public function markArchivesAsInvalidated(array $idSites, $dates, $period)
+    public function markArchivesAsInvalidated(array $idSites, array $dates, $period, Segment $segment = null, $cascadeDown = false)
     {
-        $dates = $this->removeDatesThatHaveBeenPurged($dates);
+        $invalidationInfo = new InvalidationResultInfo();
```
Comment (tsteur) [inline core/Archive/ArchiveInvalidator.php]: Can the trailing `Info` from the class name be removed maybe to have `InvalidationResult`? `Info` is a very generic term and not really meaningful as everything is a "info". `InvalidationResult` is already very meaningful. It's not important though, would be only something tiny in case you get to work on it again anyway

File: core/CronArchive.php [hunk 8]
```diff
@@ -262,6 +267,8 @@ public function __construct($processNewSegmentsFrom = null, LoggerInterface $log
 
         $processNewSegmentsFrom = $processNewSegmentsFrom ?: StaticContainer::get('ini.General.process_new_segments_from');
         $this->segmentArchivingRequestUrlProvider = new SegmentArchivingRequestUrlProvider($processNewSegmentsFrom);
+
+        $this->invalidator = StaticContainer::get('Piwik\Archive\ArchiveInvalidator');
```
Comment (tsteur) [inline core/CronArchive.php]: Was wondering why this is here and not in the method `invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain` but I guess this is to prepare later to remove the StaticContainer later easier? Should we try to do this usually? I usually do this in the method itself but if it's useful I will try to move to `__constrcut` when it makes sense

File: tests/PHPUnit/Unit/DataAccess/ArchiveTableCreatorTest.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        ArchiveTableCreator::$tablesAlreadyInstalled = $this->tables;
+
+        $tables = ArchiveTableCreator::getTablesArchivesInstalled($type);
+
+        $this->assertEquals($expectedTables, $tables);
+    }
+
+    public function getTestDataForGetTablesArchivesInstalled()
+    {
+        return array(
```
Comment (tsteur) [inline tests/PHPUnit/Unit/DataAccess/ArchiveTableCreatorTest.php]: Can you maybe add a test case for empty `$type` (or call method without params)? Would be nice to have all lines that include regex covered by unit tests

File: core/DataAccess/Model.php [hunk 10]
```diff
@@ -95,39 +98,49 @@ public function getInvalidatedArchiveIdsSafeToDelete($archiveTable, array $idSit
...
     {
-        $sql = $bind = array();
-        $datesToDelete = array_unique($datesToDelete);
-        foreach ($datesToDelete as $dateToDelete) {
-            $sql[] = '(date1 <= ? AND ? <= date2 AND name LIKE \'done%\')';
-            $bind[] = $dateToDelete;
-            $bind[] = $dateToDelete;
-        }
-        $sql = implode(" OR ", $sql);
+        $idSites = array_map('intval', $idSites);
```
Comment (tsteur) [inline core/DataAccess/Model.php]: Maybe `Common::getSqlStringFieldsArray($idSites)` would be more explicit to prevent SQL injection? Works though too so we can definitely leave it like that. Maybe a private method that contains `$idSites = array_map('intval', $idSites); implode(", ", $idSites)` could be nice just to have a description/comment what it is doing and to make it clear it's about preventing SQL injection. Not needed though

File: core/Period.php [hunk 11]
```diff
@@ -278,6 +279,26 @@ public function __toString()
...
+     * the value type of the collection. So a range period will return the result of the
+     * subperiod's `getImmediateChildPeriodLabel()` method.
+     *
+     * @ignore
+     * @return string|null
+     */
+    abstract public function getImmediateChildPeriodLabel();
+
+    /**
+     * TODO
```
Comment (tsteur) [inline core/Period.php]: Can the todo be removed maybe or a comment added?

File: tests/PHPUnit/Unit/PeriodTest.php [hunk 12]
```diff
@@ -123,4 +123,145 @@ public function getInvalidDateFormats()
...
+
+    public function getTestDataForGetAllParentPeriods()
+    {
+        return array(
+            array(
+                'day',
+                '2015-01-01',
+                array(
+                    array('week', '2014-12-29,2015-01-04'),
+                    array('month', '2014-12-01,2014-12-31'),
```
Comment (tsteur) [inline tests/PHPUnit/Unit/PeriodTest.php]: How does it come that it is valid for December?

File: tests/PHPUnit/Unit/PeriodTest.php [hunk 13]
```diff
@@ -123,4 +123,145 @@ public function getInvalidDateFormats()
...
+                    array('month', '2015-01-01,2015-01-31'),
+                    array('year', '2015-01-01,2015-12-31'),
+                ),
+            ),
+
+            array(
+                'week',
+                '2015-09-01',
+                array(
+                    array('month', '2015-08-01,2015-08-31'),
```
Comment (tsteur) [inline tests/PHPUnit/Unit/PeriodTest.php]: Also here should it be actually valid for August?

**Comment by tsteur** [comment]
> related: New plugin: provide simple UI to invalidate old reports #8942

Should we maybe put the command `core:invalidate-report-data` + tests already in that plugin so we don't have to break anything later? Or is the command too useful to have it in a plugin on the Marketplace?

**Comment by tsteur** [comment]
Left a few comments but looks good to me

File: core/CronArchive.php (hunk 8 above)
Comment (diosmosis) [inline core/CronArchive.php]: It's closer to the constructor injection pattern, but the hard part in removing StaticContainer isn't moving the code to the constructor, but making injected classes immutable. So more habit for me at this point.

File: core/DataAccess/Model.php (hunk 10 above)
Comment (diosmosis) [inline core/DataAccess/Model.php]: It's actually not just for SQL injection, but because it's much faster to not use so many placeholders when there are many sites (ie >= thousands). This is something you mentioned on slack which I remembered ;)

File: core/Period.php (hunk 11 above)
Comment (diosmosis) [inline core/Period.php]: Oops

File: tests/PHPUnit/Unit/PeriodTest.php (hunk 12 above)
Comment (diosmosis) [inline tests/PHPUnit/Unit/PeriodTest.php]: The parent periods for a day are week, month, year. The week for 2015-01-01 is 2014-12-18,2015-01-03. There are two months touched by that week period.

I think I can remove this method though, there's a less complicated approach.

**Comment by diosmosis** [comment]
> Should we maybe put the command core:invalidate-report-data + tests already in that plugin so we don't have to break anything later? Or is the command too useful to have it in a plugin on the Marketplace? 

@mattab can you answer this?

File: tests/PHPUnit/Unit/DataAccess/ArchiveTableCreatorTest.php (hunk 9 above)
Comment (diosmosis) [inline tests/PHPUnit/Unit/DataAccess/ArchiveTableCreatorTest.php]: Added

**Comment by mattab** [comment]
> Should we maybe put the command core:invalidate-report-data + tests already in that plugin so we don't have to break anything later? Or is the command too useful to have it in a plugin on the Marketplace? 

Command is too useful to have in a plugin, it must be in core for now.

File: core/DataAccess/Model.php (hunk 10 above)
Comment (tsteur) [inline core/DataAccess/Model.php]: Sweet :)

**Comment by mattab** [comment]
Updated FAQ, added:
 * Alternatively you may use the `core:invalidate-report-data` console commmand:

   ` ./console core:invalidate-report-data --dates=2012-01-01,2011-10-15 --sites=1,3,5 --dry-run`

 * when invalidating report for a period eg. `week` it automatically invalidates all periods that include this one, eg. the enclosing `month` and `year`. We say the invalidation cascades up. By default, periods that are included are not invalidated (invalidating month will not re-process the days and weeks within the month). You may also force Piwik to invalidate inner periods by setting the parameter `&cascadeDown=1`.
